yaksh/models.py

```diff
@@ -2047,20 +2048,70 @@ def get_question_answers(self):
             Return a dictionary with keys as questions and a list of the
             corresponding answers.
         """
-        q_a = {}
+        question_answers = {}
         for answer in self.answers.all():
             question = answer.question
-            if question in q_a:
-                q_a[question].append({
+            if question in question_answers:
+                question_answers[question].append({
                     'answer': answer,
                     'error_list': [e for e in json.loads(answer.error)]
                 })
             else:
-                q_a[question] = [{
+                question_answers[question] = [{
                     'answer': answer,
                     'error_list': [e for e in json.loads(answer.error)]
                 }]
-        return q_a
+        return question_answers
+
+    def get_diff_of_consecutive_submissions_for_all_code_question(self):
+        """
+            Return diff of consecutive submissions for all questions
+        """
+        question_answers = {q:self.answers.filter(question=q) for q in self.questions.all() if q.type=='code'} # key : question, value : all submitted answers for this question
+        q_diff = {} # key : question, value : list of diff between consecutive submission for this question
+        first_submissions = []
+        for question in question_answers: 
+            answers = question_answers[question]
+            first_submission = answers[0].answer
+            all_submissions_for_a_question = [] #all answers for a question as list of lines
+            for ans in answers:
+                split_ans_line_by_line = [ele for ele in ans.answer.split("\n")]
+                all_submissions_for_a_question.append(split_ans_line_by_line)
+
+            
+            if len(all_submissions_for_a_question)>1:#if there are more than one submissions for this question
+                zipped = zip(all_submissions_for_a_question, all_submissions_for_a_question[1:])
+                temp = []
+                for idx, consecutive_submissions in enumerate(zipped):
+                    diff_of_consecutive_submissions =  difflib.ndiff(consecutive_submissions[0], consecutive_submissions[1])
+                    #remove unwanted ? produced by difflib.ndiff()
+                    temp.append([ele for ele in diff_of_consecutive_submissions if not ele.startswith("?")])
+                q_diff[question] = (first_submission, zip(temp, all_submissions_for_a_question[1:]))
+
+        idx1 =1
+        idx2 =1
+        q_all_diff = {}
+        
+        for idx3, question in enumerate(q_diff):
```

Comment:
I believe you are appending combinations of all answers into one structure and passing it to the Template. This is very unscalable. You need to only send information that is requested.

Ideally the user should send information selecting which attempts he wants to diff and then you will only fetch those answers and diff them and respond with the output.

Reply:
I am combining information required to show diffs between consecutive submission for each question.